tools/resource-gen/main.go

```diff
@@ -19,6 +19,126 @@ import (
 	_ "github.com/kumahq/kuma/api/mesh/v1alpha1"
 )
 
+// NativeApuTemplate for creating a Kuma resource.
+var NativeApiTemplate = template.Must(template.New("resource").Parse(`
+// Generated by tools/resource-gen
+// Run "make generate" to update this file.
+
+{{ $pkg := printf "%s_proto" .Package }}
+{{ $tk := "` + "`" + `" }}
+
+// nolint:whitespace
+package v1alpha1
+
+import (
+	"github.com/golang/protobuf/proto"
+
+	{{ $pkg }} "github.com/kumahq/kuma/api/{{ .Package }}/v1alpha1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/kumahq/kuma/pkg/plugins/resources/k8s/native/pkg/model"
+	"github.com/kumahq/kuma/pkg/plugins/resources/k8s/native/pkg/registry"
+
+)
+
+{{range .Resources}}
+{{- if not .SkipNativeGenerate }}
+
+// +kubebuilder:object:root=true
+type {{.ResourceType}} struct {
+	metav1.TypeMeta   {{ $tk }}json:",inline"{{ $tk }}
+	metav1.ObjectMeta {{ $tk }}json:"metadata,omitempty"{{ $tk }}
+
+	Mesh   string                        {{ $tk }}json:"mesh,omitempty"{{ $tk }}
+{{- if eq .ResourceType "DataplaneInsight" }}
+	Status   *{{ $pkg }}.{{.ProtoType}} {{ $tk }}json:"status,omitempty"{{ $tk }}
+{{- else}}
+	Spec   *{{ $pkg }}.{{.ProtoType}} {{ $tk }}json:"spec,omitempty"{{ $tk }}
+{{- end}}
+}
+
+// +kubebuilder:object:root=true
+type {{.ResourceType}}List struct {
+	metav1.TypeMeta {{ $tk }}json:",inline"{{ $tk }}
+	metav1.ListMeta {{ $tk }}json:"metadata,omitempty"{{ $tk }}
+	Items           []{{.ProtoType}} {{ $tk }}json:"items"{{ $tk }}
+}
+
+{{- if not .SkipRegistration}}
+func init() {
+	SchemeBuilder.Register(&{{.ResourceType}}{}, &{{.ResourceType}}List{})
+}
+{{- end}}
+
+
+func (cb *{{.ResourceType}}) GetObjectMeta() *metav1.ObjectMeta {
+	return &cb.ObjectMeta
+}
+
+func (cb *{{.ResourceType}}) SetObjectMeta(m *metav1.ObjectMeta) {
+	cb.ObjectMeta = *m
+}
+
+func (cb *{{.ResourceType}}) GetMesh() string {
+	return cb.Mesh
+}
+
+func (cb *{{.ResourceType}}) SetMesh(mesh string) {
+	cb.Mesh = mesh
+}
+
+func (cb *{{.ResourceType}}) GetSpec() proto.Message {
+{{- if eq .ResourceType "DataplaneInsight" }}
+	return cb.Status
+{{- else}}
+	return cb.Spec
+{{- end}}
+}
+
+func (cb *{{.ResourceType}}) SetSpec(spec proto.Message) {
+{{- if eq .ResourceType "DataplaneInsight" }}
+	cb.Status = proto.Clone(spec).(*{{ $pkg }}.{{.ProtoType}})
+{{- else}}
+	cb.Spec = proto.Clone(spec).(*{{ $pkg }}.{{.ProtoType}})
+{{- end}}
+}
+
+func (cb *{{.ResourceType}}) Scope() model.Scope {
+{{- if .ScopeNamespace }}
+	return model.ScopeNamespace
+{{- else }}
+	return model.ScopeCluster
+{{- end}}
+}
+
```

Comment:
Why don't you also generate the JSON marshaller and deep copy implementations?

Reply:
Will add that in a subsequent PR.